### ISSUE TITLE
Calculation of gradient of 2D convolution operation through GradientTape returns None

### ISSUE THREAD

**Issue by kindalime** [tensorflow/tensorflow#50661]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Mint 20.1
- TensorFlow installed from (source or binary): conda
- TensorFlow version (use command below): 2.4.1
- Python version: 3.9.4
- CUDA/cuDNN version: 10.1.243, 7.6.5 (both through conda)
- GPU model and memory: GeForce RTX 2060 Rev. A, 6GB

**Describe the current behavior**
Currently, I am trying to use `GradientTape` to capture the gradient of a 2D convolution operation `tf.nn.conv2d` on a test matrix. However, when I go to actually fetch the gradients, `tape.gradient()` returns `None` instead of the expected gradient. I tested the exact same code, except I disabled eager execution and didn't use `GradientTape`, and it returned the gradients just fine.

**Describe the expected behavior**
When using `tf.nn.conv2d` inside of `GradientTape`, TensorFlow should successfully calculate the gradients.

**Standalone code to reproduce the issue**
This code works:
```
import os
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'
import numpy as np
import tensorflow as tf
tf.compat.v1.disable_eager_execution()

# sizes, fixed strides, in_channel, out_channel be 1 for now
x_size = 4
w_size = 3  # use an odd number here
x_shape = (1, x_size, x_size, 1)
w_shape = (w_size, w_size, 1, 1)
out_shape = (1, x_size - w_size + 1, x_size - w_size + 1, 1)
strides = (1, 1, 1, 1)

# numpy value
x_np = np.random.randint(10, size=x_shape)
w_np = np.random.randint(10, size=w_shape)
out_scale_np = np.random.randint(10, size=out_shape)

# tf forward
x = tf.constant(x_np, dtype=tf.float32)
w = tf.constant(w_np, dtype=tf.float32)
out = tf.nn.conv2d(input=x, filters=w, strides=strides, padding='VALID')
out_scale = tf.constant(out_scale_np, dtype=tf.float32)
f = tf.reduce_sum(tf.multiply(out, out_scale))

# tf backward
d_out = tf.gradients(f, out)[0]
d_x = tf.gradients(f, x)[0]
d_w = tf.gradients(f, w)[0]

print(d_out, d_x, d_w)
```

This code does not:
```
import os
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'
import numpy as np
import tensorflow as tf

# sizes, fixed strides, in_channel, out_channel be 1 for now
with tf.GradientTape() as g:
    x_size = 4
    w_size = 3  # use an odd number here
    x_shape = (1, x_size, x_size, 1)
    w_shape = (w_size, w_size, 1, 1)
    out_shape = (1, x_size - w_size + 1, x_size - w_size + 1, 1)
    strides = (1, 1, 1, 1)

    # numpy value
    x_np = np.random.randint(10, size=x_shape)
    w_np = np.random.randint(10, size=w_shape)
    out_scale_np = np.random.randint(10, size=out_shape)

    # tf forward
    x = tf.constant(x_np, dtype=tf.float32)
    w = tf.constant(w_np, dtype=tf.float32)
    out = tf.nn.conv2d(input=x, filters=w, strides=strides, padding='VALID')
    out_scale = tf.constant(out_scale_np, dtype=tf.float32)
    f = tf.reduce_sum(tf.multiply(out, out_scale))

# tf backward
d_out = g.gradient(f, out)[0]
d_x = g.gradient(f, x)[0]
d_w = g.gradient(f, w)[0]

print(d_out, d_x, d_w)
```

**Comment by saikumarchalla** [comment]
Was  able to reproduce the issue on colab with TF 2.5. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/b5134f90248252be82c92792ba891a2a/untitled.ipynb). Thanks!

**Comment by hasan7n** [comment]
'GradientTape' is different than `tf.gradients`.
The problem in the second code snippet is that your tensors are not being "watched" by `GradientTape`.
`tf.Variable`s are automatically watched by `GradientTape`, but tensors are not. You must "watch" them, or use `tf.Variable`s instead.
Information can be found in this guide:
[https://www.tensorflow.org/guide/eager#advanced_automatic_differentiation_topics](url)

And a note: use `persistent=True` to be able to call `g.gradient` more than one time, otherwise it will throw an error. See the code below.

So you have two options to edit your code with `GradientTape`:
option 1: watch your tensors.


    import os
    os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'
    import numpy as np
    import tensorflow as tf

    # sizes, fixed strides, in_channel, out_channel be 1 for now
    with tf.GradientTape(persistent=True) as g: # use persistent=True to be able to call g.gradient more than once
        x_size = 4
        w_size = 3  # use an odd number here
        x_shape = (1, x_size, x_size, 1)
        w_shape = (w_size, w_size, 1, 1)
        out_shape = (1, x_size - w_size + 1, x_size - w_size + 1, 1)
        strides = (1, 1, 1, 1)

        # numpy value
        x_np = np.random.randint(10, size=x_shape)
        w_np = np.random.randint(10, size=w_shape)
        out_scale_np = np.random.randint(10, size=out_shape)

        # tf forward
        x = tf.constant(x_np, dtype=tf.float32)
        w = tf.constant(w_np, dtype=tf.float32)

        #########################################################
        #########################################################
        # watch your tensors of interest
        g.watch(x)
        g.watch(w)
        #########################################################
        #########################################################

        out = tf.nn.conv2d(input=x, filters=w, strides=strides, padding='VALID')
        out_scale = tf.constant(out_scale_np, dtype=tf.float32)
        f = tf.reduce_sum(tf.multiply(out, out_scale))

    # tf backward
    d_out = g.gradient(f, out)[0]
    d_x = g.gradient(f, x)[0]
    d_w = g.gradient(f, w)[0]

    print(d_out, d_x, d_w)


option 2: use `tf.Variable`s.


    import os
    os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'
    import numpy as np
    import tensorflow as tf

    # sizes, fixed strides, in_channel, out_channel be 1 for now
    with tf.GradientTape(persistent=True) as g: # use persistent=True to be able to call g.gradient more than once
        x_size = 4
        w_size = 3  # use an odd number here
        x_shape = (1, x_size, x_size, 1)
        w_shape = (w_size, w_size, 1, 1)
        out_shape = (1, x_size - w_size + 1, x_size - w_size + 1, 1)
        strides = (1, 1, 1, 1)

        # numpy value
        x_np = np.random.randint(10, size=x_shape)
        w_np = np.random.randint(10, size=w_shape)
        out_scale_np = np.random.randint(10, size=out_shape)

        ################### EDIT ###################
        # tf forward
        x = tf.Variable(x_np, dtype=tf.float32)
        w = tf.Variable(w_np, dtype=tf.float32)
        ################### EDIT ###################

        out = tf.nn.conv2d(input=x, filters=w, strides=strides, padding='VALID')
        out_scale = tf.constant(out_scale_np, dtype=tf.float32)
        f = tf.reduce_sum(tf.multiply(out, out_scale))

    # tf backward
    d_out = g.gradient(f, out)[0]
    d_x = g.gradient(f, x)[0]
    d_w = g.gradient(f, w)[0]

    print(d_out, d_x, d_w)


`Warning`: when computing gradients using `tf.GradientTape().gradient` with respect to one variable only, the gradient is returned as a tensor and not as a list of a single tensor; so in this code you should write the final three lines without indexing, like this:

    # tf backward
    d_out = g.gradient(f, out)
    d_x = g.gradient(f, x)
    d_w = g.gradient(f, w)

**Comment by Saduf2019** [comment]
@kindalime 
You may also [refer](https://stackoverflow.com/questions/53953099/what-is-the-purpose-of-the-tensorflow-gradient-tape) to these links for better understanding,[this guide](https://www.tensorflow.org/guide/autodiff) would clear all doubts.
